### ISSUE TITLE
Lock is not released at Kernel.require when Gem.try_activate throws RuntimeException

### ISSUE THREAD

**Issue by frsyuki** [jruby/jruby#3652]
# Issues

* (a) `require` defined at `stdlib/rubygems/core_ext/kernel_require.rb` locks a monitor but doesn't release it if `Gem.try_active` throws a RuntimeException.
* (b) It throws RuntimeException at `Gem.active` with "can't add a new key into hash during iteration" message if another thread is iterating `Gem.loaded_specs`.

# Details

My application is using JRuby. I found that many threads are blocked at the same method (full stacktrace is at https://gist.github.com/frsyuki/17601f3b1483adbd6019#file-sigdump-3892-log ):

```
Thread #<Thread:0x33eb4afe> status=sleep priority=0
    uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/monitor.rb:185:in `lock'
    uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/monitor.rb:185:in `mon_enter'
    uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:39:in `require'
    uri:classloader:/embulk/guess_plugin.rb:89:in `guess'
    uri:classloader:/embulk/guess_plugin.rb:24:in `guess'
  In Java "guess" Id=100 WAITING on java.util.concurrent.locks.ReentrantLock$NonfairSync@6d71dcba owned by "qtp54440148-21" Id=21
      at sun.misc.Unsafe.park(Native Method)
      -  waiting on java.util.concurrent.locks.ReentrantLock$NonfairSync@6d71dcba
      at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
      at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
      at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireInterruptibly(AbstractQueuedSynchronizer.java:897)
      at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireInterruptibly(AbstractQueuedSynchronizer.java:1222)
      at java.util.concurrent.locks.ReentrantLock.lockInterruptibly(ReentrantLock.java:335)
      at org.jruby.RubyThread.lockInterruptibly(RubyThread.java:1844)
      at org.jruby.ext.thread.Mutex.lock(Mutex.java:91)
      ...
```

I got this Ruby stacktrace using [sigdump](https://github.com/frsyuki/sigdump).

Then, I found following error happened 6 hours ago (full stacktrace is at https://gist.github.com/frsyuki/17601f3b1483adbd6019#file-error1-log-L70 )

```
...
Caused by: org.jruby.exceptions.RaiseException: (RuntimeError) can't add a new key into hash during iteration
	at org.jruby.RubyHash.[]=(org/jruby/RubyHash.java:1006)
	at RUBY.activate(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems/specification.rb:1285)
	at RUBY.try_activate(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems.rb:196)
	at uri_3a_classloader_3a_.META_minus_INF.jruby_dot_home.lib.ruby.stdlib.rubygems.core_ext.kernel_require.require(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:126)
	at RUBY.require_and_show(uri:classloader:/embulk/plugin_registry.rb:79)
	at RUBY.search(uri:classloader:/embulk/plugin_registry.rb:39)
	at uri_3a_classloader_3a_.embulk.plugin_registry.lookup(uri:classloader:/embulk/plugin_registry.rb:26)
	at uri_3a_classloader_3a_.embulk.plugin.lookup(uri:classloader:/embulk/plugin.rb:197)
	at RUBY.new_java_parser(uri:classloader:/embulk/plugin.rb:167)
	at RUBY.new_java_parser(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/forwardable.rb:183)
```

This error happened at [lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:126](https://github.com/jruby/jruby/blob/95ab70ecb41b35898bc9fe01ace8a691db92ce32/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb#L123-L126). It assumes that `Gem.try_activate` doesn't throw exception and doesn't release acquired lock if an error happened.

I'm not sure which thread was iterating `Gem.loaded_specs` object. But it seems a public API which could be called by arbitrary code.

# Environment

* JRuby version: 9.0.0.0
* Java version: 1.8.0_51 64-Bit Server VM

**Comment by frsyuki** [comment]
I think that improving lock code at `require` method to be more defensive will solve deadlock. But it still causes RuntimeError. Serializing all access to `Gem.loaded_specs` object, or public API of it returns copy of `@loaded_spec` probably solves the entire problem but I don't know that is a possible. Any ideas are there?

**Comment by headius** [comment]
So is this a bug in RubyGems then? The locks in question are in `kernel_require` and that's not something we modify; it's just stock RubyGems. Perhaps this needs to be reported to them.

It appears that in a very recent RubyGems on our `ruby-2.3` branch, it's still not quite right.

**Comment by headius** [comment]
Here's a patch I think will work. @frsyuki if you could test this patch, all that's left is reporting to RubyGems and getting a release we can use.

```diff
diff --git a/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb b/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb
index f9b67ea..accbf53 100644
--- a/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb
+++ b/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb
@@ -120,14 +120,17 @@ module Kernel
   rescue LoadError => load_error
     RUBYGEMS_ACTIVATION_MONITOR.enter
 
-    if load_error.message.start_with?("Could not find") or
-        (load_error.message.end_with?(path) and Gem.try_activate(path)) then
-      RUBYGEMS_ACTIVATION_MONITOR.exit
-      return gem_original_require(path)
-    else
+    begin
+      if load_error.message.start_with?("Could not find") or
+          (load_error.message.end_with?(path) and Gem.try_activate(path)) then
+        require_again = true
+      end
+    ensure
       RUBYGEMS_ACTIVATION_MONITOR.exit
     end
 
+    return gem_original_require(path) if require_again
+
     raise load_error
   end
```

**Comment by frsyuki** [comment]
@headius thank you. I'll test that patch.

**Comment by frsyuki** [comment]
@headius I tried above patch for 1 week with @muga. Deadlock is not happening at least for now. Thank you!
I believe we can say that this patch solves the deadlock problem. So, the worst case is fixed but I think it's not 100% fix because `require` fails once.
Should I move this discussion to rubygems community, or CRuby maybe?

**Comment by headius** [comment]
I think we should move this discussion to a RubyGems issue. This appears to be their bug.

**Comment by headius** [comment]
I have filed rubygems/rubygems#1538. We'll leave this open until we can update 1.7 and 9k to the fixed version.

**Comment by headius** [comment]
Fixed by RG 2.6.2 in fb71835. See rubygems/rubygems#1538 for details.
